File: codebreaker/universe.py
import sys
from collections import deque

n, k = map(int,input().split())
adj = [[] for _ in range(n+1)]
cat = [0]+[int(e) for e in input().split()]
for i in range(n-1):
    u,v = map(int,input().split())
    adj[u].append(v)
    adj[v].append(u)

vis = [False for _ in range(n+1)]
con = [0 for _ in range(n+1)]
con[1] = cat[1]

vis[1] = True
cnt = 0
q = deque()
q.append(1)
while (len(q)!=0):
    u = q[0]
    q.popleft()
    if (u!=1 and len(adj[u])==1):
        cnt+=1
    for v in adj[u]:
        if vis[v]:
            continue
        if con[u]+cat[v] > k:
            continue
        if cat[v]:
            con[v]=con[u]+cat[v]
        else:
            con[v] = 0
        vis[v]=True
        q.append(v)
print(cnt)

File: codebreaker/universe.cpp
#include<bits/stdc++.h>
using namespace std;
#define int long long

int32_t main()
{
    cin.tie(0); ios::sync_with_stdio(0);
    int n,k; cin>>n>>k;
    vector<int> cat(n+1);
    for(int i=1;i<=n;i++)cin>>cat[i];
    vector<vector<int>>adj(n+1);
    vector<bool> vis(n+1,false);
    vector<int> con(n+1, 0);
    con[1] = cat[1];

    for(int i=1;i<=n-1;i++)
    {
        int u,v; cin>>u>>v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vis[1]=true;
    queue<int> q;
    int cnt=0;
    q.push(1);
    while(!q.empty())
    {
        int u = q.front(); q.pop();
        if (u!=1 and adj[u].size()==1LL) {cnt++; continue;}
        for(int v: adj[u])
        {
            if(vis[v]) continue;
            if(con[u]+cat[v]>k) continue;

            if(cat[v])con[v]=con[u]+cat[v];
            else con[v]=0;
            vis[v]=true;
            q.push(v);
        }
    }
    cout<<cnt;
    return 0;
}